File: ChatGPT/Intermediárias/longest_substring/longest_substring.cpp
#include <iostream>
#include <unordered_map>
#include <string>
using namespace std;

int lengthOfLongestSubstring(string s) {
    unordered_map<char, int> charIndexMap;
    int maxLength = 0;
    int start = 0;  // Início da janela deslizante

    for (int end = 0; end < s.length(); end++) {
        char currentChar = s[end];

        // Se o caractere está no mapa e está dentro da janela atual
        if (charIndexMap.find(currentChar) != charIndexMap.end() && charIndexMap[currentChar] >= start) {
            start = charIndexMap[currentChar] + 1;  // Mova o início da janela para a direita do último índice deste caractere
        }

        charIndexMap[currentChar] = end;  // Atualize o último índice deste caractere
        maxLength = max(maxLength, end - start + 1);  // Atualize o comprimento máximo
    }

    return maxLength;
}

int main() {
    string input = "abcabcbb";
    cout << "Length of the longest substring without repeating characters: " << lengthOfLongestSubstring(input) << endl;
    return 0;
}

File: Gemini/Intermediárias/longest_substring/longest_substring.py
def length_of_longest_substring(s: str) -> int:
    """
    Finds the length of the longest substring without repeating characters.

    Args:
        s: The input string.

    Returns:
        The length of the longest substring.
    """

    char_index_map = {}
    max_len = start = 0
    for end in range(len(s)):
        if s[end] in char_index_map and start <= char_index_map[s[end]]:
            start = char_index_map[s[end]] + 1
        char_index_map[s[end]] = end
        max_len = max(max_len, end - start + 1)
    return max_len

# Test cases
assert length_of_longest_substring("abcabcbb") == 3
assert length_of_longest_substring("bbbbb") == 1
assert length_of_longest_substring("pwwkew") == 3
